lib/srv/discovery/discovery_test.go

```diff
@@ -991,26 +990,49 @@ func TestDiscoveryServer(t *testing.T) {
 						require.Equal(t, expectedValue, got.IntegrationDiscoveredResources[expectedKey])
 					}
 					return true
-				}, 500*time.Millisecond, 50*time.Millisecond)
+				}, 1*time.Second, 50*time.Millisecond)
 			}
 			if tc.userTasksDiscoverCheck != nil {
-				var allUserTasks []*usertasksv1.UserTask
-				var nextToken string
-				for {
-					var userTasks []*usertasksv1.UserTask
-					userTasks, nextToken, err = tlsServer.Auth().UserTasks.ListUserTasks(context.Background(), 0, "", &usertasksv1.ListUserTasksFilters{})
-					require.NoError(t, err)
-					allUserTasks = append(allUserTasks, userTasks...)
-					if nextToken == "" {
-						break
-					}
-				}
-				tc.userTasksDiscoverCheck(t, allUserTasks)
+				tc.userTasksDiscoverCheck(t, tlsServer.Auth().UserTasks)
 			}
 		})
 	}
 }
 
+func fetchAllUserTasks(t *testing.T, userTasksClt services.UserTasks, minUserTasks, minUserTaskResources int) []*usertasksv1.UserTask {
+	var existingTasks []*usertasksv1.UserTask
+	require.Eventually(t, func() bool {
+		var allTasks []*usertasksv1.UserTask
+		var nextToken string
+		for {
+			var userTasks []*usertasksv1.UserTask
+			userTasks, nextTokenResp, err := userTasksClt.ListUserTasks(context.Background(), 0, nextToken, &usertasksv1.ListUserTasksFilters{})
+			require.NoError(t, err)
```

Comment:
This code runs in a background goroutine, and its not correct to fail the test outside of the test's main goroutine.

You probably want to use `EventuallyWithT` here.

Reply:
It's ok to fail the test from a different goroutine (as long as the test hasn't finished yet) but `t.FailNow` i.e. `t.Fail` followed by `runtime.Goexit` will likely not do what we wanted or expected.

How does `EventuallyWithT` help here tho?

Reply:
> FailNow must be called from the goroutine running the test or benchmark function, not from other goroutines created during the test. 

I would do something like this instead:

```
require.EventuallyWithT(t, func(t *assert.CollectT) {
    var allTasks, nextToken
    // ....
    for {
        gotResources := 0
        for _, task := range allTasks {
            gotResources += ...
        }
        assert.GreaterOrEqual(t, gotResources, minUserTasks) 
    }
})
```

I suppose `EventuallyWithT` isn't strictly required in this case, but in general when you're looking for something to happen _eventually_, you want an error to mean try again (up to the timeout), not fail the test immediately.